src/mumble/API_v_1_x_x.cpp

```diff
@@ -919,31 +919,82 @@ void MumbleAPI::getPositionalAudioData_v_1_3_x(mumble_plugin_id_t callerID,
 
 	const PositionalData &posData = pluginManager->getPositionalData();
 
-	PositionalDataNoQt positionalDataNoQt = PositionalDataNoQt();
-
 	for (int i = 0; i < 3; i++) {
 		Coord coord = static_cast< Coord >(i);
 
-		positionalDataNoQt.m_playerPos[i]  = posData.getPlayerPos()[coord];
-		positionalDataNoQt.m_playerDir[i]  = posData.getPlayerDir()[coord];
-		positionalDataNoQt.m_playerAxis[i] = posData.getPlayerAxis()[coord];
+		outPositionalData->m_playerPos[i]  = posData.getPlayerPos()[coord];
+		outPositionalData->m_playerDir[i]  = posData.getPlayerDir()[coord];
+		outPositionalData->m_playerAxis[i] = posData.getPlayerAxis()[coord];
+
+		outPositionalData->m_cameraPos[i]  = posData.getCameraPos()[coord];
+		outPositionalData->m_cameraDir[i]  = posData.getCameraDir()[coord];
+		outPositionalData->m_cameraAxis[i] = posData.getCameraAxis()[coord];
+	}
+
+	EXIT_WITH(MUMBLE_STATUS_OK);
+}
+
+void MumbleAPI::getPositionalContext_v_1_3_x(mumble_plugin_id_t callerID, char **outContext,
+									std::shared_ptr< api_promise_t > promise) {
+	if (QThread::currentThread() != thread()) {
+		// Invoke in main thread
+		QMetaObject::invokeMethod(this, "getPositionalContext_v_1_3_x", Qt::QueuedConnection,
+								  Q_ARG(mumble_plugin_id_t, callerID), Q_ARG(char **, outContext),
+								  Q_ARG(std::shared_ptr< api_promise_t >, promise));
+		return;
+	}
+
+	api_promise_t::lock_guard_t guard = promise->lock();
+	if (promise->isCancelled()) {
+		return;
+	}
+
+	VERIFY_PLUGIN_ID(callerID);
 
-		positionalDataNoQt.m_cameraPos[i]  = posData.getCameraPos()[coord];
-		positionalDataNoQt.m_cameraDir[i]  = posData.getCameraDir()[coord];
-		positionalDataNoQt.m_cameraAxis[i] = posData.getCameraAxis()[coord];
+	PluginManager *pluginManager = Global::get().pluginManager;
+	if (!pluginManager) {
+		EXIT_WITH(MUMBLE_EC_INTERNAL_ERROR);
 	}
 
-	char *context = posData.getContext().toUtf8().data();
-	size_t size = strlen(context) + 1;
-	positionalDataNoQt.m_context = static_cast< char * >(malloc(size));
-	std::memcpy(positionalDataNoQt.m_context, context, size);
+	const PositionalData &posData = pluginManager->getPositionalData();
+	const QString contextQString = posData.getContext();
+
+	char *context = contextQString.toUtf8().data();
+
+	std::memcpy(*outContext, context, static_cast<size_t>(contextQString.size()));
```

Comment:
If `outContext` happens to be smaller than `context`, this will write over and beyond the bound of the `outContext` buffer.

Also, the UTF-8 encoded version of the string in `context` will have a different size than the original UTF-16 encoded `contextQString` hence you can't use one size as a substitute for the other.

Reply:
There is no guaranteed null byte in string classes (per se). They track their size explicitly and hence don't depend on any null bytes.

Just cache the object you get from `toUtf8()` and then query it for its size.

Reply:
> There is no guaranteed null byte in string classes (per se).

Oh I didn't mean in the string class, I meant in the context string in specific, the context always has the prefix followed by a `\0` followed by the data, but yeah, your method is better